core/src/main/java/com/radixdlt/p2p/addressbook/AddressBook.java

```diff
@@ -144,9 +143,18 @@ void resetFailures(RadixNodeUri uri) {
   private final AddressBookPersistence persistence;
   private final Object lock = new Object();
   private final Map<NodeId, AddressBookEntry> knownPeers = new ConcurrentHashMap<>();
+  private int numStoredAddresses;
   private final PeerAddressEntryComparator addressEntryComparator =
       new PeerAddressEntryComparator();
 
+  /**
+   * Stores nodes whose addresses were explicitly requested (via `bestKnownAddressesById`). The
```

Comment:
I've finished my PR and come back and I'm still not sure I understand this.

I think there's a link I'm missing.
* Why is `bestKnownAddressesById` relevant? What does it mean? When is it called?
* Why should I prioritise these in my address book? What's the motivation?

Is eg a validator id discovered in a proof, so a node tries to sync from it - so the node attempts a connection to this node, which ends up hitting `bestKnownAddressesById`? And because the node was interested in connecting to this other node, it's probably worth remembering (or something?). If so, I'm wondering if putting it in `bestKnownAddressesById` is the right name/abstraction?

Reply:
> Is eg a validator id discovered in a proof, so a node tries to sync from it - so the node attempts a connection to this node, which ends up hitting bestKnownAddressesById? And because the node was interested in connecting to this other node, it's probably worth remembering (or something?).

Yup, exactly that.
PeerManager/PeerDiscovery use two AddressBook's methods to init outbound connections:
`bestCandidatesToConnect` - "I just want to connect to _some_ peers"
 `bestKnownAddressesById` - "I want to connect to this specific peer"
 
And I agree that the abstraction is not quite right :) I'll see what I can do about it (perhaps a separate method/callback that PeerManager can use directly).